opentelemetry-sdk/src/opentelemetry/sdk/metrics/__init__.py

```diff
@@ -361,6 +361,14 @@ def __init__(
         self.metrics_lock = threading.Lock()
         self.observers_lock = threading.Lock()
         self.view_manager = ViewManager()
+        self.instrument_names = set()
```

Comment:
We already have `metrics` and `observers` sets. I would change these to dict of instrument name onto observer/metric and use that instead of keeping names in a separate set. Then you can use the same locks too.

Reply:
I am wondering if we need two different containers and locks one for metric and one for observer? Is there any specific reason not to use one container and lock for both?

Reply:
@lzchen would know

I think it's to keep the instruments partitioned conveniently when the observer callbacks are run? Might not be necessary

Reply:
It was a separation because the collection of metric instrument and observers behave differently (one uses views, one simply just calls the callback). We could potentially just use one container and type check, I don't have a big opinion on this.